% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/quantgen.R
\name{estimGenRel}
\alias{estimGenRel}
\title{Genomic relatedness}
\usage{
estimGenRel(
  X,
  afs = NULL,
  thresh = NULL,
  relationships = "additive",
  method = "vanraden1",
  theta = 0.5,
  alpha = NULL,
  verbose = 1
)
}
\arguments{
\item{X}{matrix of bi-allelic SNP genotypes encoded in allele doses in \{0,1,2\}, with genotypes in rows and SNPs in columns; missing values should be encoded as NA}

\item{afs}{vector of allele frequencies, corresponding to the alleles whose copies are counted in \code{X} (if NULL, will be calculated with \code{\link{estimSnpAf}})}

\item{thresh}{threshold on minor allele frequencies below which SNPs are ignored (e.g. 0.01; NULL to skip this step)}

\item{relationships}{relationship to estimate (additive/dominant/gaussian) where "gaussian" corresponds to the Gaussian kernel from \href{http://dx.doi.org/10.3835/plantgenome2011.08.0024}{Endelman (2011)}}

\item{method}{\itemize{
\item if additive relationships, can be "noia" (see \href{https://doi.org/10.1534/genetics.116.199406}{Vitezica et al, 2017}), "vanraden1" (first method in \href{http://dx.doi.org/10.3168/jds.2007-0980}{VanRaden, 2008}), "toro2011_eq10" (equation 10 using molecular covariance from \href{http://dx.doi.org/10.1186/1297-9686-43-27}{Toro et al, 2011}), "habier" (similar to "vanraden1" but without centering; from \href{http://dx.doi.org/10.1534/genetics.107.081190}{Habier et al, 2007}), "astle-balding" (two times equation 2.2 in \href{http://dx.doi.org/10.1214/09-sts307}{Astle & Balding, 2009}), "yang" (similar to 'astle-balding' but without ignoring sampling error per SNP; from \href{http://dx.doi.org/10.1038/ng.608}{Yang et al, 2010}), "zhou" (centering the genotypes with \code{\link{scale}} and not assuming that rare variants have larger effects; from \href{http://dx.doi.org/10.1371/journal.pgen.1003264}{Zhou et al, 2013}) or "center-std";
\item if dominant relationships, can be "noia" (see \href{https://doi.org/10.1534/genetics.116.199406}{Vitezica et al, 2017}), "vitezica" (classical/statistical parametrization from \href{http://dx.doi.org/10.1534/genetics.113.155176}{Vitezica et al, 2013}) or "su" (from \href{http://dx.doi.org/10.1371/journal.pone.0045293}{Su et al, 2012})
}}

\item{theta}{smoothing parameter for "gauss"}

\item{alpha}{if not NULL, weight to make the matrix invertible as it often is singular because two genotypes are clones or because the centered-coding of SNP genotypes makes the last row predictable from the other ones (see \href{https://doi.org/10.1186/1297-9686-43-25}{Stranden and Christensen, 2011}); otherwise use \code{\link[Matrix]{nearPD}}}

\item{verbose}{verbosity level (0/1)}
}
\value{
symmetric matrix with the number of SNPs used as an attribute
}
\description{
Estimate genetic relationships between genotypes from their SNP genotypes.
Note that this function estimates "relationships" and not "coancestries".
See \href{http://dx.doi.org/10.1186/1297-9686-43-27}{Toro et al (2011)}: "for diploid individuals, twice the coancestry coefficient is the additive relationship coefficient, which describes the ratio between the genetic covariance between individuals and the genetic variance of the base population".
}
\examples{
\dontrun{set.seed(1859)
nb.genos <- 200
Ne <- 10^4
chrom.len <- 10^5
mu <- 10^(-8)
c <- 10^(-8)
genomes <- simulCoalescent(nb.inds=nb.genos,
                           pop.mut.rate=4 * Ne * mu * chrom.len,
                           pop.recomb.rate=4 * Ne * c * chrom.len,
                           chrom.len=chrom.len)
X <- genomes$genos

A.vr <- estimGenRel(X=X, relationships="additive", method="vanraden1")
mean(diag(A.vr)) # should be 1 in a base population at HWE
mean(A.vr[upper.tri(A.vr)]) # should be 0 in a base population at HWE

D.v <- estimGenRel(X=X, relationships="dominant", method="vitezica")
mean(diag(D.v)) # should be 1 in a base population at HWE
mean(D.v[upper.tri(D.v)]) # should be 0 in a base population at HWE
}
}
\author{
Timothee Flutre
}
