% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/quantgen.R
\name{imputeGenosWithMean}
\alias{imputeGenosWithMean}
\title{Genotype imputation}
\usage{
imputeGenosWithMean(X, min.maf = 0.1, max.miss = 0.3, rm.still.miss = TRUE)
}
\arguments{
\item{X}{matrix of bi-allelic SNP genotypes encoded in allele doses in \{0,1,2\}, with genotypes in rows and SNPs in columns; missing values should be encoded as NA}

\item{min.maf}{minimum minor allele frequency (before imputation) below which SNPs are discarded}

\item{max.miss}{maximum amount of missing genotypes (before imputation) above which SNPs are discarded}

\item{rm.still.miss}{if TRUE, remove SNP(s) still with missing genotype(s) after imputation (depending on \code{min.maf} and \code{max.miss})}
}
\value{
matrix with imputed genotypes
}
\description{
Impute missing SNP genotypes with the mean of the non-missing.
The code was inspired from the \code{A.mat} function from the \href{https://cran.r-project.org/web/packages/rrBLUP/}{rrBLUP} package.
}
\seealso{
\code{\link{imputeGenosWithMeanPerPop}}
}
\author{
Timothee Flutre
}
