% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/quantgen.R
\name{reformatGenoClasses}
\alias{reformatGenoClasses}
\title{Convert genotypes}
\usage{
reformatGenoClasses(
  file = NULL,
  x = NULL,
  na.string = "--",
  sep = "",
  verbose = 1
)
}
\arguments{
\item{file}{the name of the file which the data are to be read from, with a header line, columns separated by a tabulation, and row names as the first column}

\item{x}{if \code{file=NULL}, data.frame of bi-allelic SNP genotypes encoded in genotypic classes, i.e. in \{AA,AB or BA,BB\}, with SNPs in rows and genotypes in columns; if it is a matrix, it will be silently transformed into a data frame}

\item{na.string}{a character to be interpreted as NA values}

\item{sep}{separator of alleles within each SNP genotype to be removed}

\item{verbose}{verbosity level (0/1)}
}
\value{
matrix with SNPs in rows and genotypes in columns
}
\description{
Reformat bi-allelic SNP genotypes encoded in genotypic classes to ease subsequent manipulations.
}
\seealso{
\code{\link{genoClasses2genoDoses}}, \code{\link{genoClasses2JoinMap}}
}
\author{
Eric Duchene [aut], Timothee Flutre [ctb]
}
